File: SenLib/Sen3/FileFixes/r2290_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class r2290_dat : FileMod {
		public string GetDescription() {
			return "Terminology fixes in Crossbell (Ex. Camp).";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x0c5fbbb90cb3459eul, 0x0d2f5b9769eb5f7dul, 0x9d87992cu));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();
			var patcher = new SenScriptPatcher(bin);

			// royal family -> Imperial family
			patcher.ReplacePartialCommand(0x188ae, 0x53, 0x188d2, 0x3, new byte[] {0x49, 0x6d, 0x70, 0x65, 0x72, 0x69});

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/r2290.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x0c5fbbb90cb3459eul, 0x0d2f5b9769eb5f7dul, 0x9d87992cu));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/r2290.dat", file) };
		}
	}
}

File: native/sen3/file_fixes/r2290_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/sen_script_patcher.h"
#include "util/hash/sha1.h"

namespace SenLib::Sen3::FileFixes::r2290_dat {
std::string_view GetDescription() {
    return "Terminology fixes in Crossbell (Ex. Camp).";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_en/r2290.dat",
            138561,
            SenPatcher::SHA1FromHexString("0c5fbbb90cb3459e0d2f5b9769eb5f7d9d87992c"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;
        SenScriptPatcher patcher(bin);

        // royal family -> Imperial family
        patcher.ReplacePartialCommand(
            0x188ae, 0x53, 0x188d2, 0x3, {{0x49, 0x6d, 0x70, 0x65, 0x72, 0x69}});

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::r2290_dat
